The DNS Server Virtual Machine crashes due to Out-of-Memory

## :white_check_mark: Checklist

<!--
Before reporting a problem please check the following. Replace the empty checkboxes [ ] below with checked ones [x] accordingly.
-->

- [x] I have searched open and closed issues for duplicates.
- [x] This is a problem observed when **deploying** a Data Safe Haven.
- [x] I can reproduce this with the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :computer: System information

- **Operating System:** macOS
- **Data Safe Haven version:** develop@ce1d985de26043880d68d803fa8f06a8659754e2

## :package: Packages

<details>
<summary>List of packages</summary>

<!--
Python: Run `hatch run pip freeze` and paste the output below
-->

```none
Paste list of packages here
```

</details>

## :no_entry_sign: Describe the problem

PR https://github.com/alan-turing-institute/data-safe-haven/pull/2500 replaced the DNS container instance with a virtual machine running Docker. In a development TRE running uninterrupted for 4 days, that VM has crashed due to lack of memory:

<img width="1595" height="683" alt="Image" src="https://github.com/user-attachments/assets/f6e4a7f6-6e63-48e8-a643-78254889e743" />

The process responsible seems to be the `unattended-upgrade` service that runs periodically. There are multiple reports that this process causes memory problems:

* https://bugs.launchpad.net/ubuntu/+source/unattended-upgrades/+bug/2029120
* https://bugs.launchpad.net/ubuntu/+source/unattended-upgrades/+bug/1979715
* https://serverfault.com/questions/1165633/unattended-upgr-causing-my-ec2-instance-to-become-unresponsive-and-terminate

At the moment, the VM size of the DNS server is hardcoded, using a `Standard_B2ats_v2` with 1 vCPU and 1GB of memory. Those resources might be insufficient. 

As a workaround, you can use the portal to restart the VM and, if not started, restart the `adguardhome` unit.

<!--
What happens now and (if relevant) an explanation of why this is incorrect

Optionally include specific tasks (if any) in the order in which they need to be done.
- [ ] Task 1
- [ ] Task 2
- [ ] Task 3

Include a clear and concise description of what the behaviour should be when this issue is closed if this is not obvious.
-->

## :deciduous_tree: Log messages

We discovered the issue when it was impossible to log into Guacamole. When checking the container logs, we found a `java.net.UnknownHostException` error:

```none
Caused by: java.net.UnknownHostException: login.microsoftonline.com
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:567)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
	at java.base/java.net.Socket.connect(Socket.java:751)
	at java.base/sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:304)
	at java.base/sun.net.NetworkClient.doConnect(NetworkClient.java:178)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:531)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:636)
	at java.base/sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:264)
	at java.base/sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:377)
	at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:193)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1252)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1138)
	at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:179)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1690)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1614)
	at java.base/java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:531)
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:307)
	at org.jose4j.http.Get.get(Get.java:81)
	at org.jose4j.jwk.HttpsJwks.refresh(HttpsJwks.java:204)
	at org.jose4j.jwk.HttpsJwks.getJsonWebKeys(HttpsJwks.java:161)
	at org.jose4j.keys.resolvers.HttpsJwksVerificationKeyResolver.resolveKey(HttpsJwksVerificationKeyResolver.java:58)
```

When logging into the DNS VM via the serial console, we found the out-of-memory error:

```none
[230891.290388] Out of memory: Killed process 141792 (unattended-upgr) total-vm:465132kB, anon-rss:148724kB, file-rss:3112kB, shmem-rss:0kB, UID:0 pgtables:552kB oom_score_adj:0
```


</details>

## :recycle: To reproduce

<!--
How can the current behaviour be reproduced? Include some steps to reproduce the problem.

Example:
- Deploy an SRE with option X set to Y
- Attempt to perform action Z
- Observe the previously mentioned failure/error message
-->

When manually running `unattended-upgrades`, we can see that it takes about 5% of the available memory (see last spike). We can also see that the DNS server VM has between 7% and 10% of  memory available during normal operations. This situation makes out-of-memory errors likely:

<img width="1427" height="594" alt="Image" src="https://github.com/user-attachments/assets/2a66cd0e-f519-4e0a-8ac5-3ff5d6685d02" />

--- a/data_safe_haven/infrastructure/programs/sre/dns_server.py
+++ b/data_safe_haven/infrastructure/programs/sre/dns_server.py
@@ -264,7 +264,7 @@ class SREDnsServerComponent(ComponentResource):
                 resource_group_name=props.resource_group_name,
                 subnet_dns=subnet_dns,
                 virtual_network=virtual_network,
-                vm_size="Standard_B2ats_v2",
+                vm_size="Standard_B2als_v2",
             ),
             tags=child_tags,
         )

